### DISCUSSION TITLE
Debugging on IntelliJ issue - Error of Air or IntelliJ?

### DISCUSSION THREAD

**Discussion by d-angelovski** [airsdk/Adobe-Runtime-Support#1693] [Q&A]
Did not know where to raise this issue, but lets give it a start here, maybe its some not correct setup I (and few colleagues) have...

We started using IntelliJ last year, and since then we all kind of noticed this issue, but seemed like its not always there. When I try for example to debug the project (Flex application for Air mobile platform), the debugger starts and works (stopping on breakpoints) only the first time when ex. IntelliJ is freshly started. After this, it varies, it can happen that it works, and then (often) it will not. Ex: app is running, trace in console is visible, but breakpoints dont work at all. 

This can be very frustrating, when build time is a minute or two, and you need to make some test scenario, to see that debugger did not stop where it supposed to.

What I tried to do, is to put a path in JAVA.HOME=  %AIR_SDK_PATH/bin/jvm.config  so that IntelliJ is using that instead of the built in one. Did not work. Then I think I see the ADL from the Air SDK, and all seems correct. But unfortunately there is not much more that I could think of.

Anyone have same issue, and maybe more info how to track this down further down the process?

Perhaps then we can come to see (if its an issue) where it comes, and report it where needed?

**Comment by 2jfw** [comment]
With IntelliJ (AIR+Flex) having the same issues that breakpoints occasionally..
- do not work (especially when you set them _after_ the application was launched and no breakpoint existed before)
- stop working out of the blue (the former red checkmark icon of the breakpoint turns into the grey "unavailable" icon; could be a timeout issue, like pausing too long on a break point(?))
- are randomly "stopped" at _immediately_ after the application launches, showing an internal callstack (will extend post with this next time it happens). At this point the application resumes without getting paused. However, _not_ manually resuming fast enough can cause all other breakpoints to not trigger at all (timeout?)

**Comment by ajwfrost** [comment]
I'm not 100% sure how IntelliJ integrates with the Flash debugger but it's a component we've not changed (and if you're using Flex, it actually comes from the Flex SDK rather than the AIR SDK). Of course, the piece we have changed is the runtime and this needs to communicate with the client debugger process, but again it's not something we've (knowingly) updated. If we did, you'd expect it to change the behaviour for all clients not just IntelliJ.

So I would be interested to know:
- is this misbehaviour 'new'? (Maybe more a question for folk seeing this who had been using IntelliJ a lot longer...)
- has anyone seen something similar with other client apps?

I'm wondering if there's any programmatic way of setting this up to test it repeatedly to try to get it to reproduce.. the only approach I can think of currently is to start outputting more details on the communication from the runtime to the client library, e.g. via Scout... but if anyone fancies building the FDB client library with more output/logging, that might also be a good approach..

thanks

**Reply by 2jfw** [reply]
> is this misbehaviour 'new'? (Maybe more a question for folk seeing this who had been using IntelliJ a lot longer...)

Definitely not new - we are encountering this since quite some time - vaguely at the beginning of 2020 maybe. I always had the feeling that this was caused by a major IntelliJ Update, from last version 2019 to first of 2020 I think (or at least somewhen around that time frame). AIR SDK did **not** change the entire time, so pretty sure its an IntelliJ issue!

**Reply by marchbold** [reply]
Must say I haven't encountered this issue, but I very rarely use the Flex SDK so potentially the issue is there somewhere? 

Might be worth listing environment details (intellij version, OS version etc) so we can isolate the issue?

**Comment by 2jfw** [comment]
> Ex: app is running, trace in console is visible, but breakpoints dont work at all.

Can you elaborate more on how the breakpoints are displayed in IntelliJ? Are they still red/checked or grey/disabled?
One common user mistake in IntelliJ may be changing code in the classes while debugging. This will shift lines and IntelliJ loses scope. I think in FlashBuilder it was possible to change code in classes with breakpoints and the breakpoints would actually always move along correctly and still get triggered whilst in IntelliJ they will move _but_ not necessarily get triggered, so you may think they are active but there is a different line of code sitting behind the breakpoint.

**Comment by 2jfw** [comment]
![grafik](https://user-images.githubusercontent.com/73781224/155294318-b2be1446-b979-43be-adca-25f256c7fd8d.png)
This is one of the "ghost" breakpoints, which gets triggered immediately after `Player connected; session starting.`  - there is actually a breakpoint on that line where it "stopped" but it actually only looks like as if it stopped there since the application is running and is _not_ paused at that time.


**OS:** Windows 10
**IDE:** IntelliJ IDEA 2021.3.2 (Ultimate Edition) Build #IU-213.6777.52, built on January 28, 2022

Like mentioned above, this occasional issue started happening at around beginning of 2020 so it is definitely not related to that specific version of IntelliJ.

**Reply by d-angelovski** [reply]
Here is some more info regards the breakpoints. I add a breakpoint at the beginning of the application, so that to know if debugger will stop, meaning I can continue working, or if not, I need to Rebuild, and sometimes perhaps restart IntelliJ, because when it breaks, there is no coming back.

This is photos of debugger when not working, what is printed, how breakpoint looks etc:
![Screenshot at Feb 23 14-15-31](https://user-images.githubusercontent.com/23104582/155319712-a85f9102-bcde-43c3-a34e-9c419e40af4c.png)
![Screenshot at Feb 23 14-15-49](https://user-images.githubusercontent.com/23104582/155319739-e4ded50c-c14a-49c7-a2d6-20f62c9cd3f9.png)
![Screenshot at Feb 23 14-16-46](https://user-images.githubusercontent.com/23104582/155319770-b5a56efe-4f5e-4a98-b579-59889cd01ffc.png)

And this is some photos when debugger is working, and have stopped on the breakpoint:
![Screenshot at Feb 23 14-25-08](https://user-images.githubusercontent.com/23104582/155319845-19e1c2fb-9e24-4063-897b-dec2c4b73960.png)
![Screenshot at Feb 23 14-25-26](https://user-images.githubusercontent.com/23104582/155319867-dfa7b46f-9d59-4a3f-ab9a-1db3de43f613.png)
![Screenshot at Feb 23 14-26-24](https://user-images.githubusercontent.com/23104582/155319888-eac63f73-b8ef-4972-865d-d46a500270a8.png)


And this is photo of debugger working, but not stopped at a breakpoint:
![Screenshot at Feb 23 14-27-11](https://user-images.githubusercontent.com/23104582/155319950-f914dd9b-dd28-4618-ac61-725c01a25da5.png)

I guess there is difference in colors when debugger not working.

**Reply by ajwfrost** [reply]
@2jfw  - interesting that the 'ghost' breakpoint you're hitting is saying in the log "Set additional breakpoints as desired, and then type 'continue'" ... that seems to be a message straight out of the FDB console. I wonder whether they've changed the integration mechanism there, or are using a different version of FDB, that is breaking when it first runs in order to allow it to set breakpoints...

**Comment by 2jfw** [comment]
![grafik](https://user-images.githubusercontent.com/73781224/155498535-d9401752-a30b-484c-b5e6-78ce9fa56f0e.png)
This is another one of those "ghost" breakpoints, this time with a callstack. App is running without being paused though IntelliJ says otherwise.
I will contact the IntelliJ support and raise an issue there or so if someone from their team can join our discussion here.

**Reply by 2jfw** [reply]
IntelliJ Ticket Reference: https://youtrack.jetbrains.com/issue/IDEA-289604

**Comment by teotigraphix** [comment]
I have had this problem in IntelliJ before and I had to do the Remove all Break Points action to get it to reset, once that is done, setting break points works again. I haven't had this happen in a long while now though.

But it all stopped happening when I stopped using the FlexSDK(for FeathersSDK) and went to the straight AS3 AIR Harmon SDK.
